optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/manager/DefaultSolverManager.java

```diff
@@ -142,43 +152,97 @@ public void solve(
             }
             newSolverTask = new SolverTask<>(problemId, solverFactory.buildSolver(), planningProblem);
             problemIdToSolverTaskMap.put(problemId, newSolverTask);
+            problemIdToEventHandlerFuturesMap.put(problemId, Collections.synchronizedList(new ArrayList<>()));
+            problemIdToOnSolvingEndedMap.put(problemId, Optional.ofNullable(onSolvingEnded));
+            problemIdToOnExceptionMap.put(problemId, Optional.ofNullable(onException));
             logger.info("A new solver task was created with problemId ({}).", problemId);
         }
 
         // TODO implement throttling
         if (onBestSolutionChangedEvent != null) {
-            newSolverTask.addEventListener(
-                    bestSolutionChangedEvent ->
-                            eventHandlerExecutorService.submit(
-                                    () -> onBestSolutionChangedEvent.accept(bestSolutionChangedEvent.getNewBestSolution())));
+            newSolverTask.addEventListener(bestSolutionChangedEvent -> {
+                Future<?> bestSolutionChangedEventFuture = eventHandlerExecutorService.submit(
+                        () -> onBestSolutionChangedEvent.accept(bestSolutionChangedEvent.getNewBestSolution()));
+                problemIdToEventHandlerFuturesMap.get(problemId).add(bestSolutionChangedEventFuture);
+            });
         }
 
-        CompletableFuture<Solution_> solverFuture = CompletableFuture.supplyAsync(newSolverTask::startSolving, solverExecutorService);
+        submitSolverTask(newSolverTask);
+    }
+
+    private void submitSolverTask(SolverTask<Solution_> solverTask) {
+        final Object problemId = solverTask.getProblemId();
+        final Consumer<Solution_> onSolvingEnded = problemIdToOnSolvingEndedMap.get(problemId).orElse(solution_ -> {
+        });
+        final Consumer<Throwable> onException = problemIdToOnExceptionMap.get(problemId).orElse(throwable -> {
+        });
```

Comment:
If you decide to get rid of the `Optional` (see above), these calls are easily replaced with `getOrDefault(problemId, ...)`.